### PR TITLE
[adapter][move] Relax object/pure arg ordering. remove return values

### PR DESCRIPTION
**Pull request by tnowacki**
- Relaxed ordering on entry points, so object and pure arguments can be intermixed
- Added new enum CallArg to support this
- Removed return values from entry points
- Removed CallResult

A lot of the input object logic was gutted as the adapter rarely needs the owned object, so it now defers to loading an object reference. 
I removed the loading of immediate dependencies for publish transactions. It will fail fairly quickly inside the VM, but I can put the eager logic back (I'm just not sure it is needed)

### CODE REVIEW

**Comment by tnowacki** [comment]
There are some gas tests that are failing, but I'm having a bit of trouble debugging them, maybe @lxfind, you will be able to help?
One example
```
---- authority::gas_tests::test_publish_gas stdout ----
thread 'authority::gas_tests::test_publish_gas' panicked at 'assertion failed: `(left == right)`
  left: `GasCostSummary { computation_cost: 112, storage_cost: 93, storage_rebate: 0 }`,
 right: `GasCostSummary { computation_cost: 359, storage_cost: 93, storage_rebate: 0 }`', sui_core/src/unit_tests/gas_tests.rs:238:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by lxfind** [comment]
We need to load the immediate dependencies for safety reasons.
Imagine two validators are receiving the same publish transaction, but one validator doesn't have one of the dependent package on-chain. Without reading the immediate dependencies during `handle_transaction` phase, both will pass and lock the objects. But then when it comes to execution, one would succeed with publish, while the other will fail (but the failed one will still charge gas, i.e. mutate the gas object). Now the two validators have inconsistent state.

**Comment by tnowacki** [comment]
> We need to load the immediate dependencies for safety reasons.

Gotcha, makes sense. And do we not need to do it recursively? Because we assume the non-immediate deps are in a sane state?

File: sui_types/src/messages.rs [hunk 1]
```diff
@@ -157,7 +130,10 @@ impl SingleTransactionKind {
         // the same shared object doesn't show up more than once in the same single
         // transaction.
         let mut used = HashSet::new();
-        if !input_objects.iter().all(|o| used.insert(o.object_id())) {
+        if input_objects.iter().any(|o| {
+            let was_newly_inserted = used.insert(o.object_id());
+            !was_newly_inserted
+        }) {
```
Comment (lxfind) [inline sui_types/src/messages.rs:136]: Why this change?

**Comment by lxfind** [comment]
> > We need to load the immediate dependencies for safety reasons.
> 
> Gotcha, makes sense. And do we not need to do it recursively? Because we assume the non-immediate deps are in a sane state? (EDIT: And I suppose a similar argument for not checking the deps of the call?)

Correct, if a validator is aware of a package, it must have successfully executed the cert that publishes that package, which means all the dependencies of that package must be on-chain. So we don't need to check non-immediate dependencies.

File: sui_types/src/messages.rs [hunk 2]
```diff
@@ -386,15 +357,12 @@ impl<S: AuthoritySignInfoTrait> TransactionEnvelope<S> {
     }
 
     pub fn contains_shared_object(&self) -> bool {
-        match &self.data.kind {
-            TransactionKind::Single(s) => s.contains_shared_object(),
-            TransactionKind::Batch(b) => b.iter().any(|kind| kind.contains_shared_object()),
-        }
+        self.shared_input_objects().peekable().peek().is_some()
```
Comment (lxfind) [inline sui_types/src/messages.rs:360]:
```suggestion
        self.shared_input_objects().next().is_some()
```

File: sui_core/src/authority/temporary_store.rs [hunk 3]
```diff
@@ -345,10 +346,14 @@ impl<S: BackingPackageStore> ModuleResolver for AuthorityTemporaryStore<S> {
...
     fn get_module(&self, module_id: &ModuleId) -> Result<Option<Vec<u8>>, Self::Error> {
         let package_id = &ObjectID::from(*module_id.address());
+        let package_obj;
         let package = match self.read_object(package_id) {
             Some(object) => object,
             None => match self.package_store.get_package(package_id)? {
-                Some(object) => object,
+                Some(object) => {
+                    package_obj = object;
+                    &package_obj
```
Comment (lxfind) [inline sui_core/src/authority/temporary_store.rs:355]: Is this the same as `&object.clone()`?

File: sui_core/src/authority.rs [hunk 4]
```diff
@@ -643,35 +653,8 @@ impl AuthorityState {
...
-                    .all(|(kind, obj_opt)| obj_opt.is_some()
-                        || to_be_published_addresses.contains(&kind.object_id()))
-            );
-        }
-        let filtered = inputs
-            .into_iter()
-            .zip(input_objects.into_iter())
-            .filter_map(|(input, object_opt)| object_opt.map(|object| (input, object)))
-            .collect::<Vec<_>>();
-
```
Comment (lxfind) [inline sui_core/src/authority.rs:672]: Why are these deleted?

**Review by lxfind — COMMENTED** [review]
Thanks! Amazing work. This is going to take a few iterations to review.

A few high-level comments:
1. Why do we change to not pass in the objects to adapter? Is it really cleaner that way? If so, perhaps add a TODO somewhere that when we construct the TemporaryStore, we should just consume the input objects to avoid unnecessary clones.
2. This PR seems to be doing many things at once, making it difficult to review.. I recommend in the future we try to break up PRs.

**Comment by tnowacki** [comment]
>Why do we change to not pass in the objects to adapter? Is it really cleaner that way? If so, perhaps add a TODO somewhere that when we construct the TemporaryStore, we should just consume the input objects to avoid unnecessary clones.

I tried cleaning up the logic that that had a count of the number of the objects, and then used `take` off of the front of the iterator. Thats when I realized that all of those objects were already inside of the `TemporaryStore`, so no TODO needed, its already the case that this is happening. 

>This PR seems to be doing many things at once, making it difficult to review.. I recommend in the future we try to break up PRs.

I apologize about that, and I still can try to do that now that it all makes a bit more sense to me. A bit of that comes from my unfamiliarity with the codebase. This PR is really the result of me just changing things and having the type checker yell, and I then fix it until it goes away.

File: sui_core/src/authority.rs (hunk 4 above)
Comment (tnowacki) [reply sui_core/src/authority.rs:672]: The logic was gone and these asserts were already done inside of the VM, but I added it back for the reasons you stated

File: sui_core/src/authority/temporary_store.rs (hunk 3 above)
Comment (tnowacki) [reply sui_core/src/authority/temporary_store.rs:355]: No, the `object` here is already owned (`object: Object`), so the clone wouldn't do anything. This is a trick for getting around the lifetimes. If you try doing `Some(object) => &object`, it complains that the reference will not live long enough. This is a trick that forces the lifetime outside of the match, thus letting it live long enough.

File: sui_types/src/messages.rs (hunk 1 above)
Comment (tnowacki) [reply sui_types/src/messages.rs:136]: I can revert it if you want (I was probably touching this more aggressively in some earlier version of the PR)

Comment (tnowacki) [reply sui_types/src/messages.rs]: Reverted

**Comment by tnowacki** [comment]
I tried to revert most of the spurious changes

**Comment by lxfind** [comment]
Adding @patrickkuo @666lcz for visibility on the changes to the gateway API.

**Comment by lxfind** [comment]
> I tried cleaning up the logic that that had a count of the number of the objects, and then used `take` off of the front of the iterator. Thats when I realized that all of those objects were already inside of the `TemporaryStore`, so no TODO needed, its already the case that this is happening.

I think we still have two copies of the objects: one in `objects_by_kind` in authority.rs, and one in temporary store. Eventually we could consider have the construction of `AuthorityTemporaryStore` to consume the objects, which is what I think needs a TODO here.

**Comment by tnowacki** [comment]
> I think we still have two copies of the objects: one in `objects_by_kind` in authority.rs, and one in temporary store. Eventually we could consider have the construction of `AuthorityTemporaryStore` to consume the objects, which is what I think needs a TODO here.

Ah, yeah there is no real work to remove the clone there, just missed it (like no rearchitecting, just removing the & and reordering a few calls)

File: sui_programmability/adapter/src/adapter.rs [hunk 5]
```diff
@@ -625,34 +585,40 @@ fn handle_transfer<
...
+        debug_assert_eq!(*old_version, new_object.version());
     }
 }
 
 pub struct TypeCheckSuccess {
     pub module_id: ModuleId,
+    // object_owner_map maps from object ID to its exclusive object owner.
+    // This map will be used for detecting circular ownership among
+    // objects, which can only happen to objects exclusively owned
+    // by objects.
```
Comment (lxfind) [inline sui_programmability/adapter/src/adapter.rs:605]: These comments belong to where we were creating `object_owner_map` from object_data

**Review by patrickkuo — APPROVED** [review]
rest and rpc_gateway changes LGTM
